src/rebar_packages.erl

```diff
@@ -393,11 +393,37 @@ rm_ws(R) ->
     ec_semver:parse(R).
 
 valid_vsn(Vsn) ->
+    TVsn = fetch_transitive_vsn(Vsn),
     %% Regepx from https://github.com/sindresorhus/semver-regex/blob/master/index.js
     SemVerRegExp = "v?(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(\\.(0|[1-9][0-9]*))?"
         "(-[0-9a-z-]+(\\.[0-9a-z-]+)*)?(\\+[0-9a-z-]+(\\.[0-9a-z-]+)*)?",
-    SupportedVersions = "^(>=?|<=?|~>|==)?\\s*" ++ SemVerRegExp ++ "$",
-    re:run(Vsn, SupportedVersions, [unicode]) =/= nomatch.
+    SupportedVersions = "^" ++ vsn_pre_pattern() ++ SemVerRegExp ++ "$",
+    re:run(TVsn, SupportedVersions, [unicode]) =/= nomatch.
+
+vsn_pre_pattern() ->
+    "(>=?|<=?|~>|==)?\\s*".
+
+fetch_transitive_vsn(Vsn) ->
+    %% for removing `and` and `or` from dependecy and getting the highest version
+    case re:split(Vsn, <<"\s*or\s*|\s*and\s*">>) of
+        [Vsn] ->
+            Vsn;
+        Vsns ->
+            highest_version(Vsns)
+    end.
+
+highest_version(Vsns) ->
+    {Highest, _} = lists:foldl(
+        fun(Vsn, {HVsn, HVsnVal} = Acc) ->
+            VsnVal = re:replace(Vsn, vsn_pre_pattern(), "", [{return, binary}]),
+            case (HVsn =:= none orelse ec_semver:gt(VsnVal, HVsnVal)) of
```

Comment:
If the pattern here was `> 1.0.0 and < 3.0.0` this will actually pick version 3.0.0 even though the boolean expression excludes it from selection (it should go from `1.0.1` to `2.x.y`'s highest available value).

This is incorrect version resolving.

Reply:
not exactly, regex will return `> 1.0.0` and `< 3.0.0` and this function will return `< 3.0.0` which will be handled in the `resolve_version_` function i think

Reply:
I thought that replacing the `vsn_pre_pattern()` with `""` would drop the `<` and then you're left with `VsnVal` being `<<"3.0.0">>`

Reply:
That it does, but the foldl accumulator is a tuple like `{"< 3.0.0", "3.0.0"}` and we return the first element of it.